File: cadwiki-nuget/cadwiki.NUnitTestRunner/ObservableTestResults.vb
Option Strict On
Option Infer Off
Option Explicit On
Imports System
Imports System.Collections.Generic
Imports cadwiki.NUnitTestRunner.TestEvidence
Imports Newtonsoft.Json

Namespace Results
    Public Class TestResult
        Public TestName As String = ""
        Public Passed As Boolean
        Public ExceptionMessage As String = ""
        Public StackTrace As New List(Of String)
        Public Evidence As New TestEvidence.Evidence

        Public Function ToStringList() As List(Of String)
            Dim strList As New List(Of String)
            strList.Add("TestName: " + TestName)
            strList.Add("Passed: " + Passed.ToString())
            strList.Add("ExceptionMessage: " + ExceptionMessage)
            strList.Add("StackTrace: ")
            For Each trace As String In StackTrace
                strList.Add(trace)
            Next
            strList.Add("Evidence")
            strList.Add("Images: ")
            For Each image As Image In Evidence.Images
                strList.Add(image.FilePath)
            Next
            Return strList
        End Function

    End Class

    Public Class ObservableTestSuiteResults
        Public TimeElapsed As String = ""
        Public TotalTests As Integer
        Public PassedTests As Integer
        Public FailedTests As Integer
        Public Messages As New List(Of String)
        Public TestResults As New List(Of TestResult)
        Public TestSuiteName As String = ""

        Public Event MessageAdded(sender As Object, e As EventArgs)

        Public Sub AddMessage(newItem As String)
            Messages.Add(newItem)
            RaiseEvent MessageAdded(Me, New EventArgs())
        End Sub

        Public Sub SetImagePathsToRelative(reportFolder As String)
            For Each tr As TestResult In TestResults
                For Each img As Image In tr.Evidence.Images
                    If (Not String.IsNullOrEmpty(img.FilePath)) Then
                        Dim relativePath As String = img.FilePath.Replace(reportFolder, "./")
                        img.RelativeFilePath = relativePath
                    End If
                Next
            Next
        End Sub

        Public Event ResultAdded(sender As Object, e As EventArgs)

        Public Sub AddResult(newItem As TestResult)
            If newItem.Passed Then
                PassedTests += 1
            Else
                FailedTests += 1
            End If
            TotalTests += 1
            TestResults.Add(newItem)
            RaiseEvent ResultAdded(Me, New EventArgs())
        End Sub

        Public Function ToJson() As String
            Dim testSuiteResult As String = JsonConvert.SerializeObject(Me, Formatting.Indented)
            Return testSuiteResult
        End Function

        Public Function ToStringList() As List(Of String)
            Dim strList As New List(Of String)
            strList.Add(TestSuiteName)
            strList.Add("Summary")
            strList.Add("Elapsed Time: " + TimeElapsed)
            strList.Add(String.Format("{0}      Total Tests:     {1}%",
                        TotalTests.ToString(),
                        (TotalTests / TotalTests * 100.0).ToString("0.##")
                        ))
            strList.Add(String.Format("{0}/{1}  Passed Tests:    {2}%",
                        PassedTests.ToString(),
                        TotalTests.ToString(),
                        (PassedTests / TotalTests * 100.0).ToString("0.##")
                        ))
            strList.Add(String.Format("{0}/{1}  Failed Tests:    {2}%",
                        FailedTests.ToString(),
                        TotalTests.ToString(),
                        (FailedTests / TotalTests * 100.0).ToString("0.##")
                        ))
            Return strList
        End Function

    End Class

End Namespace

File: cadwiki-nuget/cadwiki.NUnitTestRunner/ObservableTestResults.cs
using System;
using System.Collections.Generic;
using cadwiki.NUnitTestRunner.TestEvidence;
using Newtonsoft.Json;

namespace cadwiki.NUnitTestRunner.Results
{
    public class TestResult
    {
        public string TestName = "";
        public bool Passed;
        public string ExceptionMessage = "";
        public List<string> StackTrace = new List<string>();
        public Evidence Evidence = new Evidence();

        public List<string> ToStringList()
        {
            var strList = new List<string>();
            strList.Add("TestName: " + TestName);
            strList.Add("Passed: " + Passed.ToString());
            strList.Add("ExceptionMessage: " + ExceptionMessage);
            strList.Add("StackTrace: ");
            foreach (string trace in StackTrace)
                strList.Add(trace);
            strList.Add("Evidence");
            strList.Add("Images: ");
            foreach (Image image in Evidence.Images)
                strList.Add(image.FilePath);
            return strList;
        }

    }

    public class ObservableTestSuiteResults
    {
        public string TimeElapsed = "";
        public int TotalTests;
        public int PassedTests;
        public int FailedTests;
        public List<string> Messages = new List<string>();
        public List<TestResult> TestResults = new List<TestResult>();
        public string TestSuiteName = "";

        public event MessageAddedEventHandler MessageAdded;

        public delegate void MessageAddedEventHandler(object sender, EventArgs e);

        public void AddMessage(string newItem)
        {
            Messages.Add(newItem);
            MessageAdded?.Invoke(this, new EventArgs());
        }

        public void SetImagePathsToRelative(string reportFolder)
        {
            foreach (TestResult tr in TestResults)
            {
                foreach (Image img in tr.Evidence.Images)
                {
                    if (!string.IsNullOrEmpty(img.FilePath))
                    {
                        string relativePath = img.FilePath.Replace(reportFolder, ".");
                        relativePath = relativePath.Replace(@"\", "/");
                        img.RelativeFilePath = relativePath;
                    }
                }
            }
        }

        public event ResultAddedEventHandler ResultAdded;

        public delegate void ResultAddedEventHandler(object sender, EventArgs e);

        public void AddResult(TestResult newItem)
        {
            if (newItem.Passed)
            {
                PassedTests += 1;
            }
            else
            {
                FailedTests += 1;
            }
            TotalTests += 1;
            TestResults.Add(newItem);
            ResultAdded?.Invoke(this, new EventArgs());
        }

        public string ToJson()
        {
            string testSuiteResult = JsonConvert.SerializeObject(this, Formatting.Indented);
            return testSuiteResult;
        }

        public List<string> ToStringList()
        {
            var strList = new List<string>();
            strList.Add(TestSuiteName);
            strList.Add("Summary");
            strList.Add("Elapsed Time: " + TimeElapsed);
            strList.Add(string.Format("{0}      Total Tests:     {1}%", TotalTests.ToString(), (TotalTests / (double)TotalTests * 100.0d).ToString("0.##")));
            strList.Add(string.Format("{0}/{1}  Passed Tests:    {2}%", PassedTests.ToString(), TotalTests.ToString(), (PassedTests / (double)TotalTests * 100.0d).ToString("0.##")));
            strList.Add(string.Format("{0}/{1}  Failed Tests:    {2}%", FailedTests.ToString(), TotalTests.ToString(), (FailedTests / (double)TotalTests * 100.0d).ToString("0.##")));
            return strList;
        }

    }

}